File: November/30/removedupeletter.java
// LeetCode 316: Remove Duplicate Letters
// Explanation:
// 1. Stack-based greedy solution
// 2. Remove larger letters that appear later
// Time Complexity: O(n)
// Space Complexity: O(26)

import java.util.*;

class Solution {
    public String removeDuplicateLetters(String s) {
        int[] count = new int[26];
        boolean[] seen = new boolean[26];
        for(char c : s.toCharArray()) count[c - 'a']++;
        Stack<Character> stack = new Stack<>();
        for(char c : s.toCharArray()){
            count[c - 'a']--;
            if(seen[c - 'a']) continue;
            while(!stack.isEmpty() && c < stack.peek() && count[stack.peek() - 'a'] > 0){
                seen[stack.pop() - 'a'] = false;
            }
            stack.push(c);
            seen[c - 'a'] = true;
        }
        StringBuilder sb = new StringBuilder();
        for(char c : stack) sb.append(c);
        return sb.toString();
    }
}

File: 2025/November/30/removedupeletter.py
# LeetCode 316: Remove Duplicate Letters
# Explanation:
# 1. Goal: Return the smallest lexicographical string with unique letters.
# 2. Use a stack to build result:
#    - If current char is smaller than top and top occurs later, pop top.
#    - Push current char if not already in stack.
# 3. Use a counter for remaining occurrences.
# Time Complexity: O(n), each character pushed/popped at most once
# Space Complexity: O(26) for stack and seen set

from collections import Counter

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        count = Counter(s)
        stack = []
        seen = set()
        for c in s:
            count[c] -= 1
            if c in seen:
                continue
            while stack and c < stack[-1] and count[stack[-1]] > 0:
                seen.remove(stack.pop())
            stack.append(c)
            seen.add(c)
        return ''.join(stack)

sol = Solution()
print(sol.removeDuplicateLetters("bcabc"))     # "abc"
print(sol.removeDuplicateLetters("cbacdcbc"))  # "acdb"
print(sol.removeDuplicateLetters("abacb"))     # "abc"
